### ISSUE TITLE
How to initialise module from inside another module?

### ISSUE THREAD

**Issue by daviddelusenet** [14islands/component-loader#5]
Hi, 

Awesome job on this library! Been using it with a lot of joy. But now I've ran into my first problem: how can I initialise a module from inside another module?

I've this code:

```
import BackgroundLoop from './BackgroundLoop';
import {Component} from 'component-loader-js';

class Background extends Component {

  constructor() {
    super(...arguments);
    this.setVariables();
  }

  setVariables() {

    this.backgroundShapes = this.el.querySelector('.js-background-shapes');
    this.shapes = [];
    this.totalSlides = 6;
    this.currentSlide = 1;
    this.prevSlide = 1;

    // create background loops
    this.backgroundLoop1 = new BackgroundLoop(this.backgroundShapes, 0, {x: 0, y: 0, color: '#1B4458'});
    this.backgroundLoop2 = new BackgroundLoop(this.backgroundShapes, 1, {x: 0, y: 0, color: '#2F637A'});

    this.backgroundLoop1.setState(1);
    this.backgroundLoop2.setState(1);

  }
}
```

As you can see I'm trying to initialise another module. This kinda works, but things like the pub/sub system break now. What is the correct way of doing this?

Thanks in advance!

**Comment by daviddelusenet** [comment]
I think I made some progress, but still everything breaks. I'm now rewriting all of my code to use the `data-component` syntax but stuff still breaks. I've this code:

```
this.shape = document.createElementNS("http://www.w3.org/2000/svg", "polygon");
this.shape.dataset.component = 'BackgroundLoopShape';
this.shape.dataset.options = '{"x":"' + nX + '", "y":"' + nY + '", "color":"' + this.getColor() + '"}' ;

// append to svg container
this.el.appendChild(this.shape);
```
This works fine. After that I try to scan the document by adding this line: `this.scan();`. This completely breaks the page, this is the error that I get:

```
Uncaught RangeError: Maximum call stack size exceeded
    at RegExp.[Symbol.match] (<anonymous>)
    at String.match (native)
    at ComponentLoader._scanElement (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:299:59)
    at http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:271:13
    at NodeList.forEach (native)
    at ComponentLoader.scan (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:270:16)
    at BackgroundLoop.scan (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:550:21)
    at BackgroundLoop.createShape (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:15847:13)
    at BackgroundLoop.setupShapes (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:15827:32)
    at new BackgroundLoop (http://127.0.0.1.xip.io:3000/static/js/hva.mdd.min.js:15778:12)
```

So the component-loader just gets into a never ending loop of initialising the same component?? How can I fix this? Please help!

**Comment by ffdead** [comment]
Hi David, happy to hear you are finding the library useful!

We've only been using it to progressively enhance components that are already defined in the markup using the `data-component` syntax. 

I think you are on the right track to define the markup dynamically and then calling `.scan()` to find the new components after injecting them in the DOM. Not sure why you get that error, but we use this process a lot when fetching new markup via ajax and then triggering a `scan()`.

However, it seems a bit overkill to use a ComponentLoader component for parts inside an SVG element. Are you sure this wouldn't be more efficient as a normal JS object/class?

**Comment by daviddelusenet** [comment]
Hi @ffdead , thanks for your reply.

I've now resorted to using the normal ES6 class definition for my SVG element paths. But this brings one problem with it: I can't use your pub/sub system. Even the solution I used before (extending your `Component` class but initialising it from inside another module by hand) breaks the pub/sub system.

Is there a way I can still use it without loading the JS component from the DOM?

**Comment by ffdead** [comment]
Ok I see, if you need pub/sub for your you child components you could pass it the mediator instance that provides the `subscribe` and `unsubscribe` methods under the hood and hook into it directly. 

It might not be the most future proof way though since this is a private member and could potentially change in future versions. You can access it in your outer component using `this.__mediator` and pass it to your child.

The component `subscribe` / `unsubscribe` methods are just proxying these methods:
```
 // Component Base Class

	/**
	 * Subscribe to an event from another component
	 * @protected
	 * @param {String} topic - Event name
	 * @param {Function} callback - Function to bind
	 */
	subscribe(topic, callback) {
		this.__mediator.subscribe(topic, callback, this);
	}


	/**
	 * Unsubscribe from an event from another component
	 * @protected
	 * @param {String} topic - Event name
	 * @param {Function} callback - Function to unbind
	 */
	unsubscribe(topic, callback) {
		this.__mediator.unsubscribe(topic, callback, this);
	}
```
But if this child component really doesn't need to communicate with other "Components" and only emit events for the parent and siblings, I would suggest using/extending something like https://github.com/primus/eventemitter3 instead.

**Comment by daviddelusenet** [comment]
Ok cool, will check this out! Thanks for all the help so far.

Now one more question: how do you handle scanning the DOM again? 

Since my `Componentloader` definition is in another file I can't access it from inside a JS component. Do you just publish an event to which the `Componentloader` listens? Or do you do something else?

**Comment by ffdead** [comment]
no problem.

This component base class has a `scan()` function you can call: https://github.com/14islands/component-loader-js/blob/master/src/component.js#L128

(as you can see the mediator is actually the shared ComponentLoader instance)

**Comment by daviddelusenet** [comment]
Yeah I know, but this just breaks for me and I don't know why. So this is the set-up I've:

```
// Component loader used for loading JavaScript components/modules
import ComponentLoader from 'component-loader-js';

// Components
import ApplyButton from 'ApplyButton';
import Background from 'Background/Background';
import BackgroundLoop from 'Background/BackgroundLoop';
import FullpageSlider from 'FullpageSlider';
import NavigationOverlay from 'NavigationOverlay';

document.addEventListener('DOMContentLoaded', function () {

  console.info("Init HvA MDD");

  const componentLoader = new ComponentLoader({
    ApplyButton,
    Background,
    BackgroundLoop,
    FullpageSlider,
    NavigationOverlay
  });

  componentLoader.scan();

}, false);
```

This works and outputs the following info in the console:

![console info](http://i.imgur.com/fVZIPUM.png)

So everything is working fine. Now let's say I add the `this.scan` line to one of my components. I'll use the `Background` component for my example. Code before:

```
class Background extends Component {

  constructor() {

    console.info('Init Background');
    super(...arguments);

    this.setVariables();

    this.setupBackgroundGradients();
    this.createShapes();
    this.transformBackground();

    this.addEventListeners();

  }

  setVariables() {

    this.backgroundShapes = this.el.querySelector('.js-background-shapes');
    this.shapes = [];
    this.totalSlides = 6;
    this.currentSlide = 1;
    this.prevSlide = 1;

  }

  // More functions

}
```

Now I add the `this.scan` line to the `setVariables` function. So this is the code after:

```
class Background extends Component {

  constructor() {

    console.info('Init Background');
    super(...arguments);

    this.setVariables();

    this.setupBackgroundGradients();
    this.createShapes();
    this.transformBackground();

    this.addEventListeners();

  }

  setVariables() {

    this.backgroundShapes = this.el.querySelector('.js-background-shapes');
    this.shapes = [];
    this.totalSlides = 6;
    this.currentSlide = 1;
    this.prevSlide = 1;
    this.scan();

  }

  // More functions

}
```

And now my console output:

![console info broken](http://i.imgur.com/Dc0ahql.png)

I didn't even edit the HTML or something, only added the `this.scan` line.

**Comment by ffdead** [comment]
Oh I see, well that's because you are calling `scan` in the constructor which creates this infinite loop.

The component is not added to the list of initialized components until the constructor has returned: 

https://github.com/14islands/component-loader-js/blob/master/src/component-loader.js#L190

When you run `scan` before the contructor has returned, the component loader detects the component again and tries to instantiate it since it's not it the list of initialized components: https://github.com/14islands/component-loader-js/blob/master/src/component-loader.js#L166

You'll need to trigger the scan later or asynchronously. Tricky use case :)

**Comment by daviddelusenet** [comment]
Makes sense, figured it was something like that. But how do I know when the constructor is returned?

Say we have something like this: componentA generates some HTML and in this HTML the `data-component="componentB"` line is found. After that HTML is generated I trigger the `this.scan` function inside of componentA but it will still break since I'm still inside componentA doing stuff.

Or is my logic just flawed? I figured something something like this would work: 

```
class Background extends Component {

  constructor() {

    console.info('Init Background');
    super(...arguments);

    this.init();

  }

  init() {

    this.setVariables();

    this.setupBackgroundGradients();
    this.createShapes();
    this.transformBackground();

    this.addEventListeners();

    this.scan();

  }
}
```

This still causes a stack overflow.

**Comment by ffdead** [comment]
Well, since you call `init()` from the constructor, it will not return until after init returns.

We haven't really used the CL in this way to dynamically add new components in the constructor, it usually happens asyncronously as the result of an Ajax call. 

In your case I wouldn't use CL components for your gradients/shapes content. But if you want to do it that way, you can force scan to run async by doing this:

```
  init() {
  ...
  setTimeout(() => scan(), 0)
  }
```

You could run into similar situations if you publish events from the constructor - depending on the order the the component loader finds your components in the markup, some of them that subscribes to that event might not have been instantiated yet. Maybe we should add a new `init()`  lifecycle method on the Component base class that is called after the scan has completed, and components would know that it's safe to trigger a new scan there or pub/sub events.

**Comment by daviddelusenet** [comment]
Cool, thanks for all the help!
